java-checks/src/main/java/org/sonar/java/symexecengine/ExecutionState.java

```diff
@@ -36,18 +37,41 @@
 
 public class ExecutionState {
 
-  final ExecutionState parent;
-  private SetMultimap<Symbol, SymbolicValue> reachableValues = HashMultimap.create();
+  public final ExecutionState parent;
+
+  /*
+   * definitelyReachableValues and potentiallyReachableValues contains the set of possible values from a given symbol.
+   * - definitelyReachableValues contains the values that are definitely retrieved by a read operation.
+   * - potentiallyReachableValues contains all the values that can be retrieved by a read operation.
+   * this is illustrated in the examples below:
+   * example 1                      example 2                        example 3
+   * a = V1;                        a = V1;                          a = V1;
+   * if(condition) {                if(condition) {                  a = V2;
+   *   a = V2;                        a = V2;
+   * } else {                       }
+   *   a = V3;
+   * }
+   * a.hashCode();                  a.hashCode();                    a.hashCode();
+   * a.unlock();                    a.unlock();                      a.unlock();
+   * definitely reachable: V2, V3   definitely reachable: V2         definitely reachable: V2
+   * potentially reachable: V2, V3  potentially reachable: V1, V2    potentially reachable: V2
+   */
+
+  // all values that are definitely modifiable (i.e. through an update).
+  private SetMultimap<Symbol, SymbolicValue> definitelyReachableValues = HashMultimap.create();
```

Comment:
both sets are points to both reachable and modifiable values:
a = new Lock(); // 1
if(condition)
  a = new Lock(); // 2
a.operation; // in case of a lock operation, this should lock 1 and 2, whereas in case of an unlock operation it should unlock only 2.

So modifiable and reachable is misleading. But I agree potentially and definitely is hard to understand so I am open to suggestion.